### PR TITLE
networkd: use BridgeFDB as well on bridge ports

### PR DESCRIPTION
**Pull request by toanju**
[BridgeFDB] did not apply to bridge ports so far. This patch adds the proper
handling. In case of a bridge interface the correct flag NTF_MASTER is now set
in the netlink call. FDB MAC addresses are now applied in
link_enter_set_addresses to make sure the link is setup.

### CODE REVIEW

File: src/network/networkd-fdb.c [hunk 1]
```diff
@@ -107,20 +107,28 @@ int fdb_entry_configure(Link *link, FdbEntry *fdb_entry) {
         _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *req = NULL;
         sd_netlink *rtnl;
         int r;
+        uint8_t flags = 0;
```
Comment (keszybz) [inline src/network/networkd-fdb.c]: No need to initialize this, since you're overwriting it a few lines below.

File: src/network/networkd-fdb.c [hunk 2]
```diff
@@ -107,20 +107,28 @@ int fdb_entry_configure(Link *link, FdbEntry *fdb_entry) {
...
         r = sd_rtnl_message_new_neigh(rtnl, &req, RTM_NEWNEIGH, link->ifindex, PF_BRIDGE);
         if (r < 0)
                 return rtnl_log_create_error(r);
 
-        /* only NTF_SELF flag supported. */
-        r = sd_rtnl_message_neigh_set_flags(req, NTF_SELF);
+        if (bridge)
+                flags |= NTF_MASTER;
+        else
+                flags |= NTF_SELF;
```
Comment (keszybz) [inline src/network/networkd-fdb.c]: Just use `=`, since flags are guaranteed to be zero so far.

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -942,6 +942,21 @@ static int link_push_ntp_to_dhcp_server(Link *link, sd_dhcp_server *s) {
...
 }
 
+static int link_set_bridge_fdb(Link *link) {
+        FdbEntry *fdb_entry;
+        int r = 0;
+
+        LIST_FOREACH(static_fdb_entries, fdb_entry, link->network->static_fdb_entries) {
+                r = fdb_entry_configure(link, fdb_entry);
+                if (r < 0) {
+                        log_link_error_errno(link, r, "Failed to add MAC entry to static MAC table: %m");
```
Comment (keszybz) [inline src/network/networkd-link.c]: `return log_link_error_errno(...);`

(Even though you're just moving the code, that's a nice cleanup.)

**Comment by toanju** [comment]
done

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -942,6 +942,20 @@ static int link_push_ntp_to_dhcp_server(Link *link, sd_dhcp_server *s) {
...
 
+static int link_set_bridge_fdb(Link *link) {
+        FdbEntry *fdb_entry;
+        int r = 0;
+
+        LIST_FOREACH(static_fdb_entries, fdb_entry, link->network->static_fdb_entries) {
+                r = fdb_entry_configure(link, fdb_entry);
+                if (r < 0) {
+                        return log_link_error_errno(link, r, "Failed to add MAC entry to static MAC table: %m");
+                }
```
Comment (poettering) [inline src/network/networkd-link.c:953]: no {} around single-line if blocks, please.

File: src/network/networkd-link.c [hunk 5]
```diff
@@ -942,6 +942,20 @@ static int link_push_ntp_to_dhcp_server(Link *link, sd_dhcp_server *s) {
         return sd_dhcp_server_set_ntp(s, addresses, n_addresses);
 }
 
+static int link_set_bridge_fdb(Link *link) {
+        FdbEntry *fdb_entry;
+        int r = 0;
```
Comment (poettering) [inline src/network/networkd-link.c:947]: i figure the = 0 assignment is redundant now.

File: src/network/networkd-link.c [hunk 6]
```diff
@@ -942,6 +942,20 @@ static int link_push_ntp_to_dhcp_server(Link *link, sd_dhcp_server *s) {
...
+
+        LIST_FOREACH(static_fdb_entries, fdb_entry, link->network->static_fdb_entries) {
+                r = fdb_entry_configure(link, fdb_entry);
+                if (r < 0) {
+                        return log_link_error_errno(link, r, "Failed to add MAC entry to static MAC table: %m");
+                }
+        }
+
+        return r;
```
Comment (poettering) [inline src/network/networkd-link.c:956]: I figure you can simplify this to `return 0;` now...

**Comment by poettering** [comment]
OK, my own expertise in these things isn't particular good, but the patch looks pretty OK. I only noticed minor coding style issues that don't really matter (see comments), hence will merge. Thanks!

**Comment by toanju** [comment]
thanks and thanks @ssahani for the coding style cleanup in #4294

**Pull request merged**